File: cpp/912_SortAnArray.cpp
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        return split(nums, 0, nums.size() - 1);
    }
    
    vector<int> split(vector<int>& nums, int start, int end) {
        if (start >= end) return {nums[start]};
        
        vector<int> first_half = split(nums, start, (start + end) >> 1);
        vector<int> second_half = split(nums, (start + end) / 2 + 1, end);
        
        return sortTwoArrays(first_half, second_half);
    }
    
    vector<int> sortTwoArrays(vector<int>& a, vector<int>& b) {
        vector<int> res;
        int left = 0, right = 0;
        while (left < a.size() || right < b.size()) {
            if (right >= b.size() || (left < a.size() && a[left] < b[right])) {
                res.push_back(a[left++]);
            } else {
                res.push_back(b[right++]);
            }
        }
        return res;
    }
};

File: java/912_SortAnArray.java
class Solution {
    public int[] sortArray(int[] nums) {
        return split(nums, 0, nums.length - 1);
    }
    
    public int[] split(int[] nums, int start, int end) {
        if (start == end) return new int[]{nums[start]};
        
        int[] firstHalf = split(nums, start, (start + end) / 2);
        int[] secondHalf = split(nums, (start + end) / 2 + 1, end);
        
        return merge(firstHalf, secondHalf);
    }
    
    public int[] merge(int[] a, int[] b) {
        int left = 0, right = 0;
        int[] mergedArray = new int[a.length + b.length];
        while (left < a.length || right < b.length) {
            if (right >= b.length || (left < a.length && a[left] < b[right])) {
                mergedArray[left + right] = a[left++];
            } else {
                mergedArray[left + right] = b[right++];
            }
        }
        return mergedArray;
    }
}
